query-algebrizer/src/clauses/predicate.rs

```diff
@@ -60,21 +71,36 @@ impl ConjoiningClauses {
         // Any variables that aren't bound by this point in the linear processing of clauses will
         // cause the application of the predicate to fail.
         let mut args = predicate.args.into_iter();
-        let left = self.resolve_numeric_argument(&predicate.operator, 0, args.next().unwrap())?;
-        let right = self.resolve_numeric_argument(&predicate.operator, 1, args.next().unwrap())?;
+        let left = args.next().expect("two args");
+        let right = args.next().expect("two args");
 
-        // These arguments must be variables or instant/numeric constants.
-        // TODO: generalize argument resolution and validation for different kinds of predicates:
-        // as we process `(< ?x 5)` we are able to check or deduce that `?x` is numeric, and either
-        // simplify the pattern or optimize the rest of the query.
-        // To do so needs a slightly more sophisticated representation of type constraints — a set,
-        // not a single `Option`.
 
+        // The types we're handling here must be the intersection of the possible types of the arguments,
+        // the known types of any variables, and the types supported by our inequality operators.
+        let left_types = self.potential_types(schema, &left)?;
+        let right_types = self.potential_types(schema, &right)?;
+        let shared_types = left_types.intersection(&right_types)
+                                     .intersection(&comparison.supported_types());
+
+        // We expect the intersection to be Long, Long+Double, Double, or Instant.
+        let left_v;
+        let right_v;
+        if shared_types == ValueTypeSet::of_one(ValueType::Instant) {
```

Comment:
I think what you mean is: can we algebrize a predicate where the types of the values are not known precisely, even if they could be mutually consistent?

I think your example needs a final clause for illustration:

```
[:find ?x ?y :where
 (or
  (and [_ :some/instant ?x]
       [_ :some/other-instant ?y])
  (and [_ :some/long ?x]
       [_ :some/other-long ?y]))
 [(< ?x ?y)]]
```

The answer is: no, at least not yet. We don't maintain possibility trees, only type sets, so the best we know about your query is that `?x` and `?y` can _each_ be _either_ `Instant` or `Long`. That means it's possible — as far as the currently implementation is concerned, anyway — for `?x` to be an `Instant` and `?y` to be a `Long`, which we're not prepared to handle.

The simplest workaround is to push the condition into each branch:


```
[:find ?x ?y :where
 (or
  (and [_ :some/instant ?x]
       [_ :some/other-instant ?y]
       [(< ?x ?y)])
  (and [_ :some/long ?x]
       [_ :some/other-long ?y]
       [(< ?x ?y)]))]
```

which will also likely be more efficient.

The other way of looking at your question is to realize that we have built ourselves this problem: the comparison operators in Clojure, and the comparisons in SQL, are overloaded. If the SQL comparison were `DATE_LESS_THAN` instead of `<`, then you wouldn't even be asking this question!